invisible_cities/detector_simulation/buffer_functions.py

```diff
@@ -0,0 +1,201 @@
+
+    stand_off = int(buffer_len * units.mus / bin_width)
+    def find_signal(wfs: pd.Series) -> List[int]:
+
+        eng_sum = wfs.sum(0)
+        indices = indices_and_wf_above_threshold(eng_sum,
+                                                 bin_threshold).indices
+        ## Just using this and the stand_off for now
+        ## taking first above sum threshold.
+        ## !! To-do: make more robust with min int? or similar
+        all_indx = split_in_peaks(indices, stand_off)
+        return [pulse[0] for pulse in all_indx]
+    return find_signal
+
+
+def padder(sensors: np.ndarray, padding: Tuple) -> np.ndarray:
+    """Pads zeros around each sensor in a 2D array"""
+    if not sensors.shape[0]:
+        return np.empty((0, padding[0] + padding[1] + 1))
+    return np.apply_along_axis(np.pad, 1, sensors, padding, "constant")
+
+
+def calculate_buffers(buffer_len: float, pre_trigger: float,
+                      pmt_binwid: float, sipm_binwid: float) -> Callable:
+    """
+    Calculates the output buffers for all sensors
+    based on a configured buffer length and pretrigger.
+    Synchronising the clock between the two sensor types.
+
+    Parameters
+    ----------
+    buffer_len  : float
+                  Length of buffer expected in mus
+    pre_trigger : float
+                  Time in buffer before identified signal in mus
+    pmt_binwid  : float
+                  Width in mus of PMT sample integration
+    sipm_binwid : float
+                  Width in mus of SiPM sample integration
+    """
+    pmt_buffer_samples  = int(buffer_len * units.mus /  pmt_binwid)
+    sipm_buffer_samples = int(buffer_len * units.mus / sipm_binwid)
+    sipm_pretrg         = int(pre_trigger * units.mus / sipm_binwid)
+    sipm_postrg         = sipm_buffer_samples - sipm_pretrg
+    pmt_pretrg_         = int(pre_trigger * units.mus / pmt_binwid)
+    pmt_postrg_         = pmt_buffer_samples - pmt_pretrg_
+
+
+    def sipm_trg_bin(sipm_bins: np.ndarray,
+                     pmt_bins : np.ndarray) -> Callable[[int], int]:
+        """
+        Identifies the SiPM bin in which the trigger
+        was identified in the PMTs.
+        Assumes finer or equal bin size for PMTs compared to SiPMs
+        """
+        def get_sipm_bin(trigger: int) -> int:
+            return np.where(sipm_bins <= pmt_bins[trigger])[0][-1]
+        return get_sipm_bin
+
+
+    def slice_generator(pmt_bins   : np.ndarray,
+                        pmt_charge : np.ndarray,
+                        sipm_bins  : np.ndarray,
+                        sipm_charge: np.ndarray) -> Generator:
+        """
+        Synchronises the clocks between the SiPMs and PMTs
+        and slices the histograms where this synchronisation
+        indicates.
+        """
+        npmt_bin  = len(pmt_bins)
+        nsipm_bin = len(sipm_bins)
+        sipm_trg  = sipm_trg_bin(sipm_bins, pmt_bins)
+
+        def generate_slices(triggers: List) -> Tuple:
+
+            for trg in triggers:
```

Comment:
Seen how you use it, I think it would be more useful to define it for a single trigger and move the loop outside.